### DISCUSSION TITLE
Why does casting to IDictionary cause problems?

### DISCUSSION THREAD

**Discussion by Ygg01** [dotnet/csharplang#7788] [Q&A] [answered]
Hi all, 

When I run this program several times on dotnet fiddle (net 6.0  and above):

```csharp
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading.Tasks;               

public class Program
{
    public static void Main()
    {
        IDictionary<string, string> dictionary = new ConcurrentDictionary<string, string>(); // FAILS sometimes
        // ConcurrentDictionary<string, string> dictionary = new ConcurrentDictionary<string, string>(); // WORKS always

        Parallel.For(0, 10, i => dictionary.TryAdd("x1", "y1"));
    }
}
```

I sometimes get `Unhandled exception. System.AggregateException: One or more errors occurred. (The key already existed in the dictionary)
System.ArgumentException: The key already existed in the dictionary.` exception

But not casting to `IDictionary` seems to make code work if not all the time, then substantially more often.

What's happening?

**Comment by HaloFour** [comment, marked as answer]
With IDictionary you're likely invoking the [TryAdd extension method](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.collectionextensions.tryadd?view=net-8.0#system-collections-generic-collectionextensions-tryadd-2(system-collections-generic-idictionary((-0-1))-0-1)) which, unless it explicitly detects ConcurrentDictionary, likely isn't a threadsafe implementation.
